### DISCUSSION TITLE
Millisecond Timer Wrap

### DISCUSSION THREAD

**Discussion by pcdiem** [arendst/Tasmota#13645] [General]
I believe the device group logic that uses the millis() function to determine if timers have elapsed may be flawed. Right now if it uses

uint32_t now = millis();
if ((long)(now - timer) >= 0)

to determine if the uint32_t timer value has been reached. According to my reading, this should handle wrapping but in my testing that doesn't seem to be true. What is the proper way to determine if a timer has past using millis() that works when millis() wraps?

**Comment by barbudor** [comment]
I've always done that and it always worked
This calculate the "distance" between the 2 times and should be working unless 64 bits arithmetic quicks in

If it actually fails, may be casting the numbers to long before the subs (instead of casting the result)

**Comment by arendst** [comment]
Look for functions ``TimeDifference``, ``TimePassedSince`` and ``TimeReached``. Theu are roll-over safe.

**Comment by pcdiem** [comment]
The logic is the same as those functions except I'm casting to long instead of int32_t. I'll change the cast to int32_t.
